pkg/metrics/metrics.go

```diff
@@ -0,0 +1,218 @@
+				Name:      "http_server_duration_seconds",
+				Help:      "The duration in seconds for outgoing server requests to be responded to.",
+				Buckets:   prometheus.LinearBuckets(.000, .05, 30),
+			},
+			[]string{"remote_address"},
+		)
+
+		oidcAuthCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: promNamespace,
+				Name:      "oidc_authentication_count",
+				Help:      "The count for OIDC authentication. Authenticated requests are 1, else 0.",
+			},
+			[]string{"authenticated", "remote_address", "user"},
+		)
+
+		tokenReviewDuration = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: promNamespace,
+				Name:      "token_review_duration_seconds",
+				Help:      "The duration in seconds for a token review lookup. Authenticated requests are 1, else 0.",
+				Buckets:   prometheus.LinearBuckets(.000, .05, 30),
+			},
+			[]string{"authenticated", "code", "remote_address", "user"},
+		)
+	)
+
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(clientRequests)
+	registry.MustRegister(clientDuration)
+	registry.MustRegister(serverRequests)
+	registry.MustRegister(serverDuration)
+	registry.MustRegister(oidcAuthCount)
+	registry.MustRegister(tokenReviewDuration)
+
+	return &Metrics{
+		registry: registry,
+
+		clientRequests: clientRequests,
+		clientDuration: clientDuration,
+
+		serverRequests: serverRequests,
+		serverDuration: serverDuration,
+
+		oidcAuthCount:       oidcAuthCount,
+		tokenReviewDuration: tokenReviewDuration,
+	}
+}
+
+// Start will register the Prometheus metrics, and start the Prometheus server
+func (m *Metrics) Start(listenAddress string) error {
+	router := http.NewServeMux()
+	router.Handle("/metrics", promhttp.HandlerFor(m.registry, promhttp.HandlerOpts{}))
+
+	ln, err := net.Listen("tcp", listenAddress)
+	if err != nil {
+		return err
+	}
+
+	m.Server = &http.Server{
+		Addr:           ln.Addr().String(),
+		ReadTimeout:    8 * time.Second,
+		WriteTimeout:   8 * time.Second,
+		MaxHeaderBytes: 1 << 15, // 1 MiB
+		Handler:        router,
+	}
+
+	go func() {
+		klog.Infof("serving metrics on %s/metrics", ln.Addr())
+
+		if err := m.Serve(ln); err != nil {
+			klog.Errorf("failed to serve prometheus metrics: %s", err)
+			return
+		}
+	}()
+
+	return nil
+}
+
+func (m *Metrics) Shutdown() error {
+	// If metrics server is not started than exit early
+	if m.Server == nil {
+		return nil
+	}
+
+	klog.Info("shutting down Prometheus metrics server...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	if err := m.Server.Shutdown(ctx); err != nil {
+		return fmt.Errorf("prometheus metrics server shutdown failed: %s", err)
+	}
+
+	klog.Info("prometheus metrics server gracefully stopped")
+
+	return nil
+}
+
+func (m *Metrics) ObserveClient(code int, path, remoteAddress string, duration time.Duration) {
+	m.clientRequests.With(prometheus.Labels{
+		"code":           strconv.Itoa(code),
+		"path":           path,
```

Comment:
What paths does the server allow? Are they static paths, or do they have dynamic parts that can take on many values?

Reply:
They are k8s paths, so all the paths you would expect to for api calls. (lots)

Reply:
Yeah, this can include things like pod names, which makes for a very large state space. I would therefore not include the path. If there is some parsing we can do on the path to extract things then it would be useful, but I'm not sure how much knowledge of the k8s api we could reasonably encode here.